app/main.py

```diff
@@ -2,114 +2,325 @@
+        except Exception as err:
+            responseError["message"] = str(err)
+            return response.json(responseError, status=400)
+
+
 @api.route("/languages")
-def get_languages(request):
+def getLanguages(request):
     """
-    Return all languages
-    Endpoint example: /languages
-    if query parameters are given, the endpoint will return that specific object
-    Example:  /languages?code=FRA
-    Parameter Name: code
+    Return all Languages objects
+    Parameters in Request body
+    {
+        "page" : (int)  [Optional, default is 1]
+    }
     """
-    queryfilter = {}
-    if len(request.args) != 0:
-        if "code" not in request.args:
-            return response.json({"message": "invalid query parameters "})
-        else:
-            queryfilter["code"] = request.args["code"][0]
+    try:
+        languages = Languages.objects()
+        args = request.json
+        if args:
+            if len(args) > 2:
+                raise Exception("Expecting less than 3 arguments")
+            if "language" in args:
+                languages = languages.filter(language=args["language"])
+            if "type" in args and "language" in args:
+                responseError["message"] = "Cannot filter with both arguments `language` and `type`"
+                return response.json(responseError, status=400)
+            if "type" in args:
+                languages = languages.filter(type=args["type"])
+            if "page" in args and int(args["page"]) > 1:
+                languages = languages.skip(int(args["page"])*ITEMS_PER_PAGE)
+                responseListObjects["page"] = args["page"]
 
-    languages = []
-    for d in my_db.languages.find(queryfilter):
-        d['_id'] = str(d['_id'])
-        languages.append(d)
-    return response.json({"message": "successfull", "data": languages})
+        responseListObjects["total_results"] = languages.count()
+        responseListObjects["total_pages"] = ceil(
+            languages.count() / ITEMS_PER_PAGE)
+        languages.limit(ITEMS_PER_PAGE)
+        responseListObjects["results"] = [d.serialize() for d in languages]
+        return response.json(responseListObjects)
+    except Exception as err:
+        responseError["message"] = err
+        return response.json(responseError, status=400)
 
 
 @api.route("/sentences")
-def get_sentences(request):
+def getSentences(request):
     """
-    return all sentences. If a parameter is given, then all sentences from that language
-    are returned
-    Endpoint example: /sentences?lang=fra
-    Query Parameter: (optional) lang 
+    Return all Sentences objects
+    Parameters in Request body
+    {
+        "page": (int)[Optional, default is 1]
+    }
     """
-    queryfilter = {}
-    if len(request.args) != 0:
-        if "lang" not in request.args:
-            return response.json({"message": "invalid query parameters "})
+    try:
+        sentences = Sentences.objects()
```

Comment:
does the ORM provide some support for pagination? If this collection has a lot of lines, billions this might crash the server. It is not super important yet as we don't expect that load.

Reply:
The ORM supports pagination. The query `.objects()` returns a cursor. If I set the limit before checking the page number, I am afraid I will always the elements from the first page. That's why I get all the lines before slicing the part the I really need.